ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/NettyHttpResponsePipeline.kt

```diff
@@ -0,0 +1,345 @@
+     */
+    private var previousCallHandled: ChannelPromise = context.newPromise().also {
+        it.setSuccess()
+    }
+
+    /**
+     * Flush if there is some unflushed data, nothing to read from channel and no active requests
+     */
+    internal fun flushIfNeeded() {
+        if (
+            isDataNotFlushed.value &&
+            httpHandler.isChannelReadCompleted.value &&
+            httpHandler.activeRequests.value == 0L
+        ) {
+            context.flush()
+            isDataNotFlushed.compareAndSet(expect = true, update = false)
+        }
+    }
+
+    internal fun processResponse(call: NettyApplicationCall) {
+        call.previousCallFinished = previousCallHandled
+        call.callFinished = context.newPromise()
+        previousCallHandled = call.callFinished
+
+        processElement(call)
+    }
+
+    private fun processElement(call: NettyApplicationCall) = processOrderedCall(call) {
+        try {
+            processCall(call)
+        } catch (actualException: Throwable) {
+            processCallFailed(call, actualException)
+        } finally {
+            call.responseWriteJob.cancel()
+        }
+    }
+
+    /**
+     * Process [call] with [block] when the response is ready and previous call is successfully processed.
+     * [call] won't be processed with [block] if previous call is failed.
+     */
+    private fun processOrderedCall(call: NettyApplicationCall, block: () -> Unit) {
+        call.response.responseReady.addListener responseFlag@{ responseFlagResult ->
+            call.previousCallFinished.addListener waitPreviousCall@{ previousCallResult ->
+                if (!previousCallResult.isSuccess) {
+                    processCallFailed(call, previousCallResult.cause())
+                    return@waitPreviousCall
+                }
+                if (!responseFlagResult.isSuccess) {
+                    processCallFailed(call, responseFlagResult.cause())
+                    return@waitPreviousCall
+                }
+                block.invoke()
+            }
+        }
+    }
+
+    private fun processCallFailed(call: NettyApplicationCall, actualException: Throwable) {
+        val t = when {
+            actualException is IOException && actualException !is ChannelIOException ->
+                ChannelWriteException(exception = actualException)
+            else -> actualException
+        }
+
+        call.response.responseChannel.cancel(t)
+        call.responseWriteJob.cancel()
+        call.response.cancel()
+        call.dispose()
+        call.callFinished.setFailure(t)
+    }
+
+    private fun processUpgradeResponse(call: NettyApplicationCall, responseMessage: Any): ChannelFuture {
+        val future = context.write(responseMessage)
+        call.upgrade(context)
+        call.isByteBufferContent = true
+
+        context.flush()
+        isDataNotFlushed.compareAndSet(expect = true, update = false)
+        return future
+    }
+
+    private fun finishCall(
```

Comment:
Please consider renaming this and writing a comment

Reply:
Is `proceedLastMessage` a good name? What do you think?

Reply:
I'm not sure about proceed